app/services/course/material/zip_download_service.rb

```diff
@@ -2,11 +2,68 @@ class Course::Material::ZipDownloadService
   class << self
     # Downloads the materials and zip them.
     #
-    # @param [Course::Material::Folder] folder The folder containing the materials.
     # @param [Array<Course::Material>] materials The materials to be downloaded.
+    # @param [Course::Material::Folder] folder The folder containing the materials. This is the
+    #   lowest common ancestor folder of all the materials.
+    # @return [String] The path to the zip file.
+    def download_and_zip(materials, folder)
+      dir = download(materials, folder)
+      zip_dir(dir)
+    end
+
+    private
+
+    # Download materials and store them in a temporary directory.
+    #
+    # @param [Array<Course::Material>] materials The materials to be downloaded.
+    # @param [Course::Material::Folder] folder The folder containing the materials.
+    # @return [String] The path to the temporary directory.
+    def download(materials, folder)
+      materials = Array(materials)
+      tmp_dir = Dir.mktmpdir('coursemology-download-')
+      Dir.chdir(tmp_dir)
```

Comment:
"If a block is given, it is passed the name of the new current directory, and the block is executed with that as the current directory. The original working directory is restored when the block exits. The return value of chdir is the value of the block. chdir blocks can be nested, but in a multi-threaded program an error will be raised if a thread attempts to open a chdir block while another thread has one open."

Seems the block way works.